File: packages/boxsh/src/tar.ts
// TAR archive encoding and decoding.
import type { StorageBackend } from "./backend.js";

const enc = new TextEncoder();
const dec = new TextDecoder();
const BLOCK = 512;

function octal(n: number, width: number): Uint8Array {
  const s = n.toString(8).padStart(width - 1, "0") + "\0";
  return enc.encode(s);
}

function header(name: string, size: number, mtime: number, isDir: boolean): Uint8Array {
  const h = new Uint8Array(BLOCK);
  const nameBytes = enc.encode(isDir ? name + "/" : name);
  if (nameBytes.length > 100) {
    // split into prefix/name at a slash boundary
    const full = isDir ? name + "/" : name;
    let cut = full.lastIndexOf("/", 100);
    if (cut <= 0) cut = 100;
    h.set(enc.encode(full.slice(cut + 1)).subarray(0, 100), 0);
    h.set(enc.encode(full.slice(0, cut)).subarray(0, 155), 345);
  } else {
    h.set(nameBytes, 0);
  }
  h.set(octal(isDir ? 0o755 : 0o644, 8), 100); // mode
  h.set(octal(0, 8), 108); // uid
  h.set(octal(0, 8), 116); // gid
  h.set(octal(isDir ? 0 : size, 12), 124);
  h.set(octal(Math.floor(mtime / 1000), 12), 136);
  h.set(enc.encode("        "), 148); // checksum placeholder = spaces
  h[156] = isDir ? 0x35 : 0x30; // typeflag '5' | '0'
  h.set(enc.encode("ustar\0"), 257);
  h.set(enc.encode("00"), 263);
  let sum = 0;
  for (const b of h) sum += b;
  h.set(enc.encode(sum.toString(8).padStart(6, "0") + "\0 "), 148);
  return h;
}

export function tarExport(backend: StorageBackend): Uint8Array {
  const parts: Uint8Array[] = [];
  const walk = (dir: string): void => {
    for (const name of (backend.list(dir) ?? []).sort()) {
      const full = dir === "" ? name : `${dir}/${name}`;
      const e = backend.entry(full);
      if (!e) continue;
      if (e.kind === "dir") {
        parts.push(header(full, 0, e.mtime, true));
        walk(full);
      } else {
        const data = backend.read(full) ?? new Uint8Array(0);
        parts.push(header(full, data.length, e.mtime, false));
        parts.push(data);
        const pad = (BLOCK - (data.length % BLOCK)) % BLOCK;
        if (pad) parts.push(new Uint8Array(pad));
      }
    }
  };
  walk("");
  parts.push(new Uint8Array(BLOCK * 2)); // end-of-archive
  const out = new Uint8Array(parts.reduce((n, p) => n + p.length, 0));
  let at = 0;
  for (const p of parts) {
    out.set(p, at);
    at += p.length;
  }
  return out;
}

function cstr(buf: Uint8Array): string {
  const end = buf.indexOf(0);
  return dec.decode(end === -1 ? buf : buf.subarray(0, end));
}

/** Import entries into the backend (merge/overwrite; missing parents created). */
export function tarImport(backend: StorageBackend, tar: Uint8Array): void {
  const mkdirs = (path: string): void => {
    const segs = path.split("/").filter(Boolean);
    let cur = "";
    for (const s of segs) {
      cur = cur === "" ? s : `${cur}/${s}`;
      if (!backend.entry(cur)) backend.mkdir(cur);
    }
  };
  let at = 0;
  while (at + BLOCK <= tar.length) {
    const h = tar.subarray(at, at + BLOCK);
    if (h.every((b) => b === 0)) break;
    const prefix = cstr(h.subarray(345, 500));
    const name = (prefix ? prefix + "/" : "") + cstr(h.subarray(0, 100));
    const size = parseInt(cstr(h.subarray(124, 136)).trim() || "0", 8);
    const type = h[156];
    at += BLOCK;
    const clean = name.replace(/\/+$/, "").replace(/^\/+/, "");
    if (type === 0x35) {
      if (clean) mkdirs(clean);
    } else if (type === 0x30 || type === 0) {
      if (clean) {
        mkdirs(clean.split("/").slice(0, -1).join("/"));
        backend.write(clean, tar.slice(at, at + size));
      }
      at += Math.ceil(size / BLOCK) * BLOCK;
    } else {
      at += Math.ceil(size / BLOCK) * BLOCK; // skip links/other types
    }
  }
}

File: crates/boxsh-fs/src/tar.rs
//! TAR archive encoding and decoding — the Rust twin of tar.ts.
//!
//! Plain ustar: names over 100 bytes split into prefix/name at a slash
//! boundary, mtimes stored at second precision, directories as type '5'
//! entries. Import merges into the backend (overwrite; missing parents
//! created), stamping mtimes with the backend clock exactly like the
//! TypeScript implementation. Links and other entry types are skipped.

use crate::{Backend, Kind, Result, split_parent};

const BLOCK: usize = 512;

fn set_octal(h: &mut [u8], at: usize, width: usize, n: u64) {
    let s = format!("{n:0>w$o}\0", w = width - 1);
    h[at..at + width].copy_from_slice(&s.as_bytes()[..width]);
}

fn header(name: &str, size: u64, mtime_ms: u64, is_dir: bool) -> [u8; BLOCK] {
    let mut h = [0u8; BLOCK];
    let full = if is_dir {
        format!("{name}/")
    } else {
        name.to_string()
    };
    let fb = full.as_bytes();
    if fb.len() > 100 {
        // Split into prefix/name at a slash boundary (mirrors tar.ts,
        // including its handling of names with no early slash).
        let cut = match fb[..fb.len().min(101)].iter().rposition(|&b| b == b'/') {
            Some(0) | None => 100,
            Some(i) => i,
        };
        let tail = &fb[(cut + 1).min(fb.len())..];
        let tail = &tail[..tail.len().min(100)];
        h[..tail.len()].copy_from_slice(tail);
        let pre = &fb[..cut.min(155)];
        h[345..345 + pre.len()].copy_from_slice(pre);
    } else {
        h[..fb.len()].copy_from_slice(fb);
    }
    set_octal(&mut h, 100, 8, if is_dir { 0o755 } else { 0o644 });
    set_octal(&mut h, 108, 8, 0); // uid
    set_octal(&mut h, 116, 8, 0); // gid
    set_octal(&mut h, 124, 12, if is_dir { 0 } else { size });
    set_octal(&mut h, 136, 12, mtime_ms / 1000);
    h[148..156].copy_from_slice(b"        "); // checksum placeholder
    h[156] = if is_dir { b'5' } else { b'0' };
    h[257..263].copy_from_slice(b"ustar\0");
    h[263..265].copy_from_slice(b"00");
    let sum: u32 = h.iter().map(|&b| u32::from(b)).sum();
    let cs = format!("{sum:0>6o}\0 ");
    h[148..156].copy_from_slice(cs.as_bytes());
    h
}

/// The whole tree as a tar archive.
pub fn export<B: Backend + ?Sized>(backend: &mut B) -> Result<Vec<u8>> {
    let mut out = Vec::new();
    walk(backend, "", &mut out)?;
    out.resize(out.len() + BLOCK * 2, 0); // end-of-archive
    Ok(out)
}

fn walk<B: Backend + ?Sized>(backend: &mut B, dir: &str, out: &mut Vec<u8>) -> Result<()> {
    let mut names = backend.list(dir)?;
    names.sort();
    for name in names {
        let full = if dir.is_empty() {
            name
        } else {
            format!("{dir}/{name}")
        };
        let Some(e) = backend.entry(&full)? else {
            continue;
        };
        match e.kind {
            Kind::Dir => {
                out.extend_from_slice(&header(&full, 0, e.mtime, true));
                walk(backend, &full, out)?;
            }
            Kind::File => {
                let data = backend.read(&full)?;
                out.extend_from_slice(&header(&full, data.len() as u64, e.mtime, false));
                out.extend_from_slice(&data);
                let pad = (BLOCK - data.len() % BLOCK) % BLOCK;
                out.resize(out.len() + pad, 0);
            }
        }
    }
    Ok(())
}

fn cstr(buf: &[u8]) -> String {
    let end = buf.iter().position(|&b| b == 0).unwrap_or(buf.len());
    String::from_utf8_lossy(&buf[..end]).into_owned()
}

fn parse_octal(field: &[u8]) -> usize {
    let s = cstr(field);
    usize::from_str_radix(s.trim(), 8).unwrap_or(0)
}

fn mkdirs<B: Backend + ?Sized>(backend: &mut B, path: &str) -> Result<()> {
    let mut cur = String::new();
    for seg in path.split('/').filter(|s| !s.is_empty()) {
        if cur.is_empty() {
            cur.push_str(seg);
        } else {
            cur = format!("{cur}/{seg}");
        }
        if backend.entry(&cur)?.is_none() {
            backend.mkdir(&cur)?;
        }
    }
    Ok(())
}

/// Import entries into the backend (merge/overwrite; missing parents created).
pub fn import<B: Backend + ?Sized>(backend: &mut B, tar: &[u8]) -> Result<()> {
    let mut at = 0usize;
    while at + BLOCK <= tar.len() {
        let h = &tar[at..at + BLOCK];
        if h.iter().all(|&b| b == 0) {
            break;
        }
        let prefix = cstr(&h[345..500]);
        let name_field = cstr(&h[..100]);
        let name = if prefix.is_empty() {
            name_field
        } else {
            format!("{prefix}/{name_field}")
        };
        let size = parse_octal(&h[124..136]);
        let typeflag = h[156];
        at += BLOCK;
        let clean = name.trim_matches('/');
        if typeflag == b'5' {
            if !clean.is_empty() {
                mkdirs(backend, clean)?;
            }
        } else if typeflag == b'0' || typeflag == 0 {
            if !clean.is_empty() {
                let (parent, _) = split_parent(clean);
                mkdirs(backend, parent)?;
                let end = (at + size).min(tar.len());
                backend.write(clean, &tar[at.min(end)..end])?;
            }
            at += size.div_ceil(BLOCK) * BLOCK;
        } else {
            at += size.div_ceil(BLOCK) * BLOCK; // skip links/other types
        }
    }
    Ok(())
}
